File: Programacao_Funcional/alta_ordem/codes/capitalize.hs
import Data.Char

capitalize [] = []
capitalize (x:xs) = cap x : capitalize xs where
    cap [] = []
    cap (y:ys) = toUpper y : lower ys
    lower [] = []
    lower (z:zs) = toLower z : lower zs

main = putStr $ unlines $ capitalize xs where
    xs = ["abc", "XYZ", "Teste", "iPod"]

File: Programacao_Funcional/alta_ordem/codes/capitalize.cpp
#include <bits/stdc++.h>

using namespace std;

vector<string> capitalize(const vector<string>& xs) {
    vector<string> ys;

    for (auto x : xs) {
        auto y = x;

        if (not y.empty()) {
            y[0] = toupper(x[0]);

            for (size_t i = 1; i < x.size(); ++i)
                y[i] = tolower(x[i]);
        }

        ys.push_back(y);
    }

    return ys;
}

int main()
{
    vector<string> xs { "abc", "XYZ", "Teste", "iPod" };

    auto ys = capitalize(xs);

    for (auto y : ys)
        cout << y << '\n';

    return 0;
}
